### PR TITLE
added FollowNewDiscussions option

### PR DESCRIPTION
**Pull request by oscarotero**
With this option, all users will be subscribed and notified for every discussion started.

### CODE REVIEW

**Comment by luceos** [comment]
Relates to https://github.com/flarum/core/issues/768

**Comment by luceos** [comment]
Very nice, someone has to test this out before merging. But really amazing job @oscarotero :+1:

**Comment by oscarotero** [comment]
Cheers :smile:
Just two things:
- I added the button text 'Follow new discussions automatically' directly but it shoud be included in the translations file.
- I noted that `DiscussionWasStarted` event is triggered after `PostWasPosted`, so I had to duplicate the code to send the notification. If the events were triggered in reverse order I could remove this code because it was send by `SendNotificationWhenReplyIsPosted` class

File: src/Listener/FollowNewDiscussions.php [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+use Flarum\Core\Access\AssertPermissionTrait;
+use Flarum\Core\User;
+use Flarum\Event\ConfigureUserPreferences;
+use Flarum\Event\DiscussionWasStarted;
+use Illuminate\Contracts\Events\Dispatcher;
+use Flarum\Subscriptions\Notification\NewPostBlueprint;
+
+class FollowNewDiscussions
+{
+    use AssertPermissionTrait;
```
Comment (franzliedke) [inline src/Listener/FollowNewDiscussions.php]: This is not used anywhere, right?

Comment (oscarotero) [inline src/Listener/FollowNewDiscussions.php]: Right. I'll remove it.

File: src/Listener/FollowNewDiscussions.php [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+    }
+
+    /**
+     * @param PostWasPosted $event
+     */
+    public function whenDiscussionWasStarted(DiscussionWasStarted $event)
+    {
+        $users = User::all()->filter(function ($user) {
+            return $user->exists && $user->getPreference('followNewDiscussions');
+        });
```
Comment (franzliedke) [inline src/Listener/FollowNewDiscussions.php:59]: Hmm, this is not a good idea. Fetching all users before filtering them will not work. The filtering has to be done in the database - which might turn out to be difficult, due to how we've stored the preferences.

Any idea, @tobscure?

File: js/forum/src/addSubscriptionSettings.js [hunk 3]
```diff
@@ -13,5 +13,13 @@ export default function() {
         onchange: this.preferenceSaver('followAfterReply')
       })
     );
+
+    items.add('followNewDiscussions',
+      Switch.component({
+        children: 'Follow new discussions automatically',
```
Comment (franzliedke) [inline js/forum/src/addSubscriptionSettings.js]: Please add a translation key here. If you have the time, please also send a PR to flarum/flarum-ext-english. (If you don't, that's fine, then we'll add it.)

Comment (oscarotero) [inline js/forum/src/addSubscriptionSettings.js]: Done: https://github.com/flarum/flarum-ext-english/pull/75

BTW: I don't think it's a good idea to save the translations of all plugins in the same repository. Each plugin should manage its own translations.

File: src/Listener/FollowNewDiscussions.php (hunk 2 above)
Comment (tobyzerner) [inline src/Listener/FollowNewDiscussions.php:59]: Quite simply, the preference needs to be stored in a new column. Or we need to change the way preferences are stored in core (a user_preferences table with columns user_id, key, and value).

Comment (luceos) [inline src/Listener/FollowNewDiscussions.php:59]: I'd say moving it to a seperate table makes a lot of sense.

**Comment by franzliedke** [comment]
@oscarotero Thanks again for this pull request!

Do you think you will have time to work on the suggested improvements?

**Comment by oscarotero** [comment]
@franzliedke I'm a bit bussy these days but I can try. Anyway, if you're having more time and want to work on this, feel free.

I like the `user_preferences` table approach, because this can be reused by other plugins and it's easier to clean plugin specific preferences on uninstall them. I know this is a more deep change that affects to database but this is the time to make breaking changes, isn't? 😃
Other option is to use the [msql json field format](https://dev.mysql.com/doc/refman/5.7/en/json.html) but I'm not familiarized with this and it's not supported by mysql 5.5 and 5.6.

**Comment by franzliedke** [comment]
@oscarotero How about this? We'll take care of the preferences refactoring, and when it's done, we let you finish this PR? :)

**Comment by oscarotero** [comment]
Fair enough. I look forward to this change.

**Comment by oscarotero** [comment]
@franzliedke I've found a security issue that I've fixed temporarily in a different branch (used in my private installation), related with the ability of the tags plugin to hide some discussions under specific tags.

As you can see here: https://github.com/oscarotero/flarum-ext-subscriptions/blob/galician-locale/src/Listener/FollowNewDiscussions.php#L67 I check whether the discussion has tags and discard the users without permissions to view discussions under any of these tags.

I don't know if there's a better way to handle this, not dependent of a specific plugin (there might be other plugins changing the view permissions). How about a more agnostic `$discussion->hasPermission('user.[id].view')`?

BTW, May this issue be present also in the other options provided by this plugin (users receiving notifications of discussions can not see?)

**Comment by jbrgr** [comment]
Hi all,

I know this is a rather old PR but it covers a feature I'm looking for. Is it actually possible to finalize this implementation or is it too outdated?

Thanks

**Comment by franzliedke** [comment]
@jbrgr There is a WIP PR that changes how we store user preferences - that's a prerequisite for finishing this one.

**Comment by jbrgr** [comment]
@franzliedke This must be https://github.com/flarum/core/pull/1467 if I'm not mistaken. As it has been rebased onto master recently it's maybe not too far away from being completed. 👍

**Comment by franzliedke** [comment]
Indeed, that's what I meant. We will review (and take care of) all open PRs for beta.9.

**Comment by franzliedke** [comment]
Closing because of follow-up PR #20.

**Pull request closed without merging**